### PR TITLE
Naive Support for Hopper FP8 Prefill Kernel with Per-Head Quantization

### PR DESCRIPTION
**Pull request by happierpig**
## Summary

This PR introduces naive FP8 tensor core computation following [FA3's implementations](https://github.com/Dao-AILab/flash-attention/blob/bdf733be55f0b323a8cf7cc6745a81c3f43cd7f0/hopper/flash_api.cpp). The main code modifications are located in `include/flashinfer/attention/hopper/quantization`, with test cases in `src/fp8-dev`. The primary changes include:

1. **In-Kernel V Transpose**  
   Since [`wgmma.fp8`](https://github.com/NVIDIA/cutlass/blob/e9627ce55b42fd2599f58cd4396da9380954def0/include/cute/arch/mma_sm90.hpp#L1774) requires K-major for both operands, and Q/K/V are all head_dim-major,  V transpose is required before feeding into tensor core. Therefore we provide an in-kernel transpose in shared memory using `ldmatrix/stmatrix` ([kernel_traits.cuh#L54](https://github.com/happierpig/flashinfer-ai/blob/4fad7efe9bc56da7400fedaf382fbfc9c362191a/include/flashinfer/attention/hopper/quantization/kernel_traits.cuh#L54))

2. **P Requantization**  
   After `Q * K` multiplication and before `P * V`, P is requantized per tensor using an oracle scale:  `p_scale = std::numeric_limits::max();`. This is based on the observation that the maximum value of P is `1` in online softmax. This strategy follows the approach in [SageAttention](https://arxiv.org/abs/2411.10958), which increases the utilization of 8-bit width compared to direct cast.
3. **Fused Dequantization**  
To reduce CUDA core computation overhead, both `QK` and `PV` dequantization steps are fused into existing online softmax operations: QK dequantization is fused into `sm_scale` ([code reference](https://github.com/happierpig/flashinfer-ai/blob/4fad7efe9bc56da7400fedaf382fbfc9c362191a/include/flashinfer/attention/hopper/variants.cuh#L73)). PV dequantization is fused into the [`finalize`](https://github.com/happierpig/flashinfer-ai/blob/4fad7efe9bc56da7400fedaf382fbfc9c362191a/include/flashinfer/attention/hopper/attention_updater.cuh#L221) step, where the denominator is applied to the output.

## Remaining Work

- [ ] Upstream modifications to separate `head_dim_qk` and `head_dim_v`
- [ ] Optimize performance to close the gap between FA3 and FlashInfer FP8, possibly tuning kernel launch parameters
- [ ] Add sparse and quantized support in `sparse_mainloop.cuh`

## Perf Benchmarks on H100
FlashInfer-FP8 on average provides 20-30% throughput boost compared to FP16. However, there exists a performance gap between FA3-FP8, calling for further optimizations. Ref to [scripts](https://github.com/happierpig/flashinfer-ai/blob/4fad7efe9bc56da7400fedaf382fbfc9c362191a/src/fp8-dev/bench_single_prefill_sm90.cu).

image
image
image

## Correctness (MSE)
To validate accuracy, we compute MSE between different FP8 implementations, and output from FP16 FlashInfer. Ref to [scripts](https://github.com/happierpig/flashinfer-ai/blob/4fad7efe9bc56da7400fedaf382fbfc9c362191a/src/fp8-dev/test_single_prefill_fa3_sm90.cu). Our impl is slightly better.
image

### CODE REVIEW

File: include/flashinfer/attention/hopper/variants.cuh [hunk 1]
```diff
@@ -61,9 +61,47 @@ struct LogitsSoftCap {
...
+};
+
 template <bool use_logits_soft_cap>
 using DefaultAttention = std::conditional_t<use_logits_soft_cap, LogitsSoftCap, StandardAttention>;
+using DefaultFP8Attention = StandardFP8Attention;
 
 }  // namespace flashinfer
 
-#endif  // FLASHINFER_ATTENTION_HOPPER_VARIANTS_CUH_
+#endif  // FLASHINFER_ATTENTION_HOPPER_VARIANTS_CUH_
```
Comment (yzh119) [inline include/flashinfer/attention/hopper/variants.cuh]: Can you use `pre-commit` to format code?

**Review by yzh119 — APPROVED** [review]
I'm good with the PR in general and let's merge it first and iterate.

File: include/flashinfer/attention/hopper/variants.cuh (hunk 1 above)
Comment (happierpig) [reply include/flashinfer/attention/hopper/variants.cuh]: Sounds good. Done.

**Review by yzh119 — APPROVED** [review]
